contracts/src/L2/L2BridgeController.sol

```diff
@@ -0,0 +1,183 @@
+     */
+    function onERC1155Received(
+        address /* operator */,
+        address from,
+        uint256 tokenId,
+        uint256 /* amount */,
+        bytes calldata data
+    ) external virtual override onlyRegistry returns (bytes4) {
+        // If from is not address(0), it's not a mint operation - process as ejection
+        if (from != address(0)) {
+            _processEjection(tokenId, data);
+        }
+        
+        return this.onERC1155Received.selector;
+    }
+
+    /**
+     * @dev Default implementation of onRenew that does nothing.
+     * Can be overridden in derived contracts for custom behavior.
+     */
+    function onRenew(uint256 tokenId, uint64 expires, address /*renewedBy*/) external virtual {
+        bridge.sendMessage(BridgeEncoder.encodeRenewal(tokenId, expires));
+    }
+
+    /**
+     * @dev Private helper method to complete name transfer operations
+     * @param tokenId The token ID of the name
+     * @param transferData The transfer data containing subregistry, resolver, and owner
+     */
+    function _completeNameTransfer(
+        uint256 tokenId,
+        TransferData memory transferData
+    ) private {
+        // owner should be the bridge controller
+        if (registry.ownerOf(tokenId) != address(this)) {
+            revert NotTokenOwner(tokenId);
+        }
+
+        registry.setSubregistry(tokenId, IRegistry(transferData.subregistry));
+        registry.setResolver(tokenId, transferData.resolver);
+
+        // now unset the token observer and transfer the name to the owner
+        registry.setTokenObserver(tokenId, ITokenObserver(address(0)));
+        registry.safeTransferFrom(address(this), transferData.owner, tokenId, 1, "");
+    }
+
+    /**
+     * Overrides the EjectionController._onEject function.
+     */
+    function _onEject(uint256[] memory tokenIds, TransferData[] memory transferDataArray) internal virtual override {
+        for (uint256 i = 0; i < tokenIds.length; i++) {
+            uint256 tokenId = tokenIds[i];
+            TransferData memory transferData = transferDataArray[i];
+
+            // check that the label matches the token id
+            _assertTokenIdMatchesLabel(tokenId, transferData.label);
+
+            /*
+            Check that there is no more than holder of the token observer and subregistry setting roles.
+
+            This works by calculating the no. of assignees for each of the given roles as a bitmap `(counts & mask)` where each role's corresponding 
+            nybble is set to its assignee count.
+
+            Since the roles themselves are bitmaps where each role's nybble is set to 1, we can simply comparing the two values to 
+            check to see if each role has exactly one assignee.
+
+            We also don't need to check that we (the bridge controller) are the sole assignee of these roles since we exercise these 
+            roles further down below.
+            */
+            uint256 roleBitmap = 
```

Comment:
Should we also prevent ejection if another account can set the resolver?

Reply:
I don't see why we need that - we deliberately don't touch the resolver so that it keeps resolving however the owner wants it to.